sound/soc/codecs/rt711-sdca-sdw.c

```diff
@@ -0,0 +1,390 @@
+	unsigned int buf;
+
+	dev_dbg(&slave->dev,
+		"%s control_port_stat=%x, sdca_cascade=%x", __func__,
+		status->control_port, status->sdca_cascade);
+
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &rt711->scp_sdca_stat1);
+	if (ret < 0)
+		goto io_error;
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &rt711->scp_sdca_stat2);
+	if (ret < 0)
+		goto io_error;
+
+	do {
+		/* clear flag */
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &buf);
+		if (ret < 0)
+			goto io_error;
+		if (buf & SDW_SCP_SDCA_INTMASK_SDCA_0) {
+			ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT1, SDW_SCP_SDCA_INTMASK_SDCA_0);
+			if (ret < 0)
+				goto io_error;
+		}
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &buf);
+		if (ret < 0)
+			goto io_error;
+		if (buf & SDW_SCP_SDCA_INTMASK_SDCA_8) {
+			ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT2, SDW_SCP_SDCA_INTMASK_SDCA_8);
+			if (ret < 0)
+				goto io_error;
+		}
+
+		/* check if flag clear or not */
+		ret = regmap_read(rt711->regmap, SDW_DP0_INT, &buf);
+		if (ret < 0)
+			goto io_error;
+		sdca_cascade = buf & SDW_DP0_SDCA_CASCADE;
+
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &buf);
+		if (ret < 0)
+			goto io_error;
+		scp_sdca_stat1 = buf & SDW_SCP_SDCA_INTMASK_SDCA_0;
+
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &buf);
+		if (ret < 0)
+			goto io_error;
+		scp_sdca_stat2 = buf & SDW_SCP_SDCA_INTMASK_SDCA_8;
+
+		stat = scp_sdca_stat1 || scp_sdca_stat2 || sdca_cascade;
+
+		count++;
+	} while (stat != 0 && count < retry);
+
+	if (stat)
+		dev_warn(&slave->dev,
+			"%s scp_sdca_stat1=0x%x, scp_sdca_stat2=0x%x\n", __func__,
+			rt711->scp_sdca_stat1, rt711->scp_sdca_stat2);
+
+	if (status->sdca_cascade)
+		mod_delayed_work(system_power_efficient_wq,
+			&rt711->jack_detect_work, msecs_to_jiffies(250));
+
+	return 0;
+
+io_error:
+	pr_err_ratelimited("IO error in %s, ret %d\n", __func__, ret);
+	return ret;
+}
+
+static struct sdw_slave_ops rt711_sdca_slave_ops = {
+	.read_prop = rt711_sdca_read_prop,
+	.interrupt_callback = rt711_sdca_interrupt_callback,
+	.update_status = rt711_sdca_update_status,
+};
+
+static int rt711_sdca_sdw_probe(struct sdw_slave *slave,
+				const struct sdw_device_id *id)
+{
+	struct regmap *regmap, *mbq_regmap;
+
+	/* Regmap Initialization */
+	mbq_regmap = regmap_init_sdw_mbq(slave, &rt711_sdca_mbq_regmap);
+	if (IS_ERR(mbq_regmap))
+		return PTR_ERR(mbq_regmap);
+
+	regmap = devm_regmap_init_sdw(slave, &rt711_sdca_regmap);
+	if (IS_ERR(regmap))
+		return PTR_ERR(regmap);
+
+	return rt711_sdca_init(&slave->dev, regmap, mbq_regmap, slave);
+}
+
+static int rt711_sdca_sdw_remove(struct sdw_slave *slave)
+{
+	struct rt711_sdca_priv *rt711 = dev_get_drvdata(&slave->dev);
+
+	if (rt711 && rt711->hw_init)
+		cancel_delayed_work(&rt711->jack_detect_work);
+
+	return 0;
+}
+
+static const struct sdw_device_id rt711_sdca_id[] = {
+	SDW_SLAVE_ENTRY_EXT(0x025d, 0x711, 0x3, 0x1, 0),
+	{},
+};
+MODULE_DEVICE_TABLE(sdw, rt711_sdca_id);
+
```

Comment:
I suddenly find "#ifdef CONFIG_PM" is missed on recent realtech sdw codec driver, but it was included, .e.g. 
 https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/codecs/rt5682.c#L2883.  Is this not necessary in kernel now ?

Reply:
static const struct dev_pm_ops rt711_sdca_pm = {
	SET_SYSTEM_SLEEP_PM_OPS(rt711_sdca_dev_suspend, rt711_sdca_dev_resume)
	SET_RUNTIME_PM_OPS(rt711_sdca_dev_suspend, rt711_sdca_dev_resume, NULL)
};

Please check the macro definitions.
https://github.com/thesofproject/linux/blob/0f9b8fc5f8705a16940299195b4ecbee29d3c229/include/linux/pm.h#L305
https://github.com/thesofproject/linux/blob/0f9b8fc5f8705a16940299195b4ecbee29d3c229/include/linux/pm.h#L341

Reply:
I understand SET_SYSTEM_SLEEP_PM_OPS, I mean using PM_CONFIG to remove rt711_sdca_dev_suspend. According to Pierre, __maybe_unused is preferred now